### PR TITLE
Integration/boosttest

### PR DESCRIPTION
**Pull request by guwirth**
* add initial files to setup an integration test for boost and boost unit tests
* currently only files without build & check

### CODE REVIEW

**Comment by wenns** [comment]
>1200 files for a sample project? Its way too big for my taste. Can we get it smaller?

**Comment by jmecosta** [comment]
Would you consider moving the integration test data into a sub module?

**Comment by guwirth** [comment]
;-> that's boost.

Boost test framework is on top of boost. I already removed a lot of directories. Next step would be to cleanup within the directories. Problem I see is that this is a lot of work each time we upgrade to a new boost version. Will try to find out touched files with sysinternal tools...

**Comment by wenns** [comment]
How about installing it via
```
$ sudo apt-get install libboost-test-dev
```

**Comment by wenns** [comment]
> Would you consider moving the integration test data into a sub module?
I dont think its necessary, at least at this point of time. I think we can manage the problem otherwise.

**Comment by guwirth** [comment]
Will try with sudo. Do you know to which location boost will be installed? Are there afterwards environment variables I can use? Someone experience with all that stuff?

**Comment by wenns** [comment]
Just to be clear: this command line would work on Ubuntu and Debian derivates and looks differently with other package managers. My point is just that: you can install boost test on the system, just like e.g. cppcheck. I.e.:
* tests, build scripts, scripts for running the tests come into the sample project
* libs, headers are installed on the system. You need it to only to develop the tests anyway. Once development is ready, you just checkin the ready XML reports and run the tests against them. 

As for locations of the libs and headers etc.: on Linux they just land in standard locations (headers in /usr/include, libs in /usr/lib) and it just works. To make building work cross platform, one can just keep (and change, if necessary) in constants inside of build files.

**Comment by guwirth** [comment]
@wenns : would like to write the tests now. 
* Should I add it to test_execution_statistics.feature or is it better to split it google-test/boost-test?
* What is the entry point? Just add a new .feature file?

**Comment by wenns** [comment]
Please add new scenarios to the test_execution_statistics.feature. Your test cases are the same feature, after all. Behave's execution procedure is like so:
* behave knows the folder where to look for features via .behaverc
* the according step implementations (the python code) is found via convention
* the python code in steps is imported, analysed and put into the library of step implementations
* the features are read in, the step names parsed, according step implementations lookuped and called

**Comment by guwirth** [comment]
@wenns tried to add a first test to *test_execution_statistics.feature* but it's not executed. See nothing in the log file on the Travis server. Do you have any idea?

File: .travis.yml [hunk 1]
```diff
@@ -4,7 +4,10 @@ jdk:
   - oraclejdk8
   - openjdk7
 
-
+  
+before_install:
+        - sudo apt-get install libboost-dev libboost-test-dev
```
Comment (jmecosta) [inline .travis.yml]: this will not work on windows. or on any non deb base system... isnt there a better way of handling these...

**Comment by guwirth** [comment]
> sudo apt-get install libboost-dev libboost-test-dev

@jmecosta: there is also the possibility
* to download the zip and extract it
* to mock the boost unit test header with some simple defines

@wenns, @jmecosta:
But maybe some word what I like to do: My main focus is to add new tests for boosttest, here I like to test the filename way and the AST-based way. To test the AST-based way I need the header files to see if the plugin is able to parse boost headers.
At the moment my main interest is to add all needed test and to bring it into operation. This boostlib thing is a priority B point which could be solved at the end.

File: .travis.yml (hunk 1 above)
Comment (wenns) [inline .travis.yml]: The .travis.yml will be executed on a Travis VM, which is garanteed to be a debian derivate ;)
Having said this, the line is not actually needed, as the behave will only run 'sonar-runner' against checked in reports, no?
The boost header, libs, etc. are needed only for developing the tests, i.e. on your workstation. Not on the CI slave.

So not probs here

wont be executed on Windows

Comment (jmecosta) [inline .travis.yml]: Of course, did not notice that was in the yaml file.. Good as it is :)

**Comment by guwirth** [comment]
> Remove 'Outline' and it will execute. 

:-< copied from the wrong example: 'Scenario: googletest report is invalid' is also wrong

> The boost header, libs, etc. are needed only for developing the tests

Still not sure about that. Like to test AST based approach for this I need the header files otherwise unit tests can't be parsed in the right way?

**Comment by wenns** [comment]
> :-< copied from the wrong example: 'Scenario: googletest report is invalid' is also wrong

Ahh. Damn. You're right!

>> The boost header, libs, etc. are needed only for developing the tests
> Still not sure about that. Like to test AST based approach for this I need the header files otherwise 
> unit  tests can't be parsed in the right way?

[Had a look into the test files] Oh yes, looks like...

**Comment by guwirth** [comment]
- adapt boost xsl to support changed xunit feature: support all log_level settings
- extend SourceCodeBuilderVisitor for functions to return qualified names including namespaces
- extend xunit lookup table: supports search for qualified and unqualified classname now
- add a new unit test to search for qualified classname with namespace
- use 'sudo apt-get install libboost-dev libboost-test-dev' for test with real boost
- add boost test examples
- add boost test reports
- add new scenarios to integration tests:
  - add 'Scenario Outline: boosttest reports cannot be found or are empty'
  - add 'Scenario: Importing nested boosttest report in default mode'
  - add 'Scenario: Importing unchanged boosttest reports in default mode'
  - add 'Scenario Outline: Importing unchanged boosttest reports in detailed mode (filename tag)'
  - add 'Scenario Outline: Importing unchanged boosttest reports in detailed mode (AST)'
  - add 'Scenario: Test with real boost test framework'

- close #401
- close #372
- close #368
- close #375

File: cxx-squid/src/main/java/org/sonar/cxx/CxxAstScanner.java [hunk 2]
```diff
@@ -124,7 +125,18 @@ public SourceCode createSourceCode(SourceCode parentSourceCode, AstNode astNode)
           sb.append(token.getValue());
         }
         String functionName = sb.toString();
-        SourceFunction function = new SourceFunction(functionName + ":" + astNode.getToken().getLine());
```
Comment (wenns) [inline cxx-squid/src/main/java/org/sonar/cxx/CxxAstScanner.java]: This ...

File: cxx-squid/src/main/java/org/sonar/cxx/CxxAstScanner.java [hunk 3]
```diff
@@ -209,4 +221,23 @@ public SourceCode createSourceCode(SourceCode parentSourceCode, AstNode astNode)
     return builder.build();
   }
 
+  // Concatenate two strings, but if there is overlap at the intersection,
```
Comment (wenns) [inline cxx-squid/src/main/java/org/sonar/cxx/CxxAstScanner.java]: ... and this looks somewhat worrisome. Can you explain the reason for doing such gymnastics here ? :)

Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/CxxAstScanner.java]: For sure ;->

If you are inside of a namespace/class it is allowed to do a definition with the namespace/class again. This function ensures that namespace/class is not doubled. 

 namespace example {
     class A {
          void f1();
          void f2();
          void f3() {} // example::A::f3
          void A::f4() {} // example::A::f4
      };
     void example::A::f1() {} // example::A::f1
     void A::f2() {} // example::A::f2
 }

File: cxx-squid/src/main/java/org/sonar/cxx/CxxAstScanner.java (hunk 2 above)
Comment (guwirth) [inline cxx-squid/src/main/java/org/sonar/cxx/CxxAstScanner.java]: This part is creating the qualified name including the namespaces. Going from current node upwards the tree adding all namespace names.

**Comment by wenns** [comment]
The tests dont run on my machine (Linux 32 bit). wth? They did pass Travis and they obviously passed on your machine. But I get a bunch of:
```
19:09:57.386 DEBUG - Performing the 'filename'-based lookup using the value 'tests/unittests/test_component1.cc'
19:09:57.386 WARN  - ... no resource found, the testcase 'my_module:foo_successfull/foo_successfull' has to be skipped
```
where the lookup is expected to succeed. What can cause this? The stuff I merged in the meanwhile? Replacement of the deprecated ModuleFileSystem? Hmm...

**Comment by guwirth** [comment]
:-< rebase to see if last changes are the issue

**Comment by guwirth** [comment]
Fails after rebase, so last changes on master are the root cause.

**Comment by guwirth** [comment]
seems to be changed behaviour in ```DefaultResourceFinder.findInSonar```?

**Comment by guwirth** [comment]
@Typz, @wenns original search in fs.testDirs() is missing now?
Bug or feature?

Old:
```Java
 public org.sonar.api.resources.File findInSonar(File file, SensorContext context, ModuleFileSystem fs, Project project) {
    org.sonar.api.resources.File unitTestFile = org.sonar.api.resources.File.fromIOFile(file, project);
    if (unitTestFile == null) {
      unitTestFile = org.sonar.api.resources.File.fromIOFile(file, fs.testDirs());
    }
    if (context.getResource(unitTestFile) == null) {
      unitTestFile = null;
    }
    return unitTestFile;
  }
```
New:
```Java
 public org.sonar.api.resources.File findInSonar(File file, SensorContext context, FileSystem fs, Project project) {
    org.sonar.api.resources.File unitTestFile = org.sonar.api.resources.File.fromIOFile(file, project);
    if (context.getResource(unitTestFile) == null) {
      unitTestFile = null;
    }
    return unitTestFile;
  }
```

**Comment by wenns** [comment]
Yep, saw that already. The point is, the older version had two access trials because of < 4.0 compatibility issue of the like... The new one should work too. Will try to investigate further.

**Comment by guwirth** [comment]
http://javadocs.sonarsource.org/4.5/apidocs/org/sonar/api/resources/File.html

public static File fromIOFile(File file,  Project module)
Deprecated. since 4.5 use FileSystem.inputFile(org.sonar.api.batch.fs.FilePredicate)

**Comment by wenns** [comment]
We're still using 4.2 API.

**Comment by guwirth** [comment]
@wenns are you looking for a solution?

**Comment by wenns** [comment]
Right now Im cleaning rooms ;)
But I want to invest an hour or two this weekend. Are you?

**Comment by guwirth** [comment]
Have a time slot between 4 and 6 o'clock. If you want I can do another try...

**Comment by wenns** [comment]
Lets make it that way: I'll just merge the stuff and skip failing tests. And we'll address the problem on the master. I mean: the PR is OK, its master thats broken. This PR just exposes a problem.

**Comment by wenns** [comment]
Pushed to master. Failing tests are marked with @skip. If you want to run them, remove this annotation.

**Pull request closed without merging**